### PR TITLE
Catch Nondep_cannot_erase

### LINKED ISSUE #9858: Compiler fails with Ctype.Nondep_cannot_erase exception.
`ocaml` (versions 4.09.0 and 4.11.0) fails on the following code:
```ocaml
module type Vector_space = sig
  type t
  type scalar
  val scale : scalar -> t -> t
end

module type Scalar = sig
  type t
  include Vector_space with type t := t
                        and type scalar = t
end

module type Linear_map = sig
  type ('a, 'b) t
  val scale :
    (module Vector_space with type t = 'a and type scalar = 'l) ->
    'l -> ('a, 'a) t
end

module Primitive(Linear_map : Linear_map) = struct
  let f (type s) ((module S : Scalar with type t = s) as s) x =
    Linear_map.scale s x
end
```
with the following exception:
```sh
$ocaml bug.ml
Fatal error: exception Ctype.Nondep_cannot_erase(_)
```
This is my best effort to produce a minimal example, but there may well still be some useless bits.

### PR DESCRIPTION
**Pull request by trefis**
Fixes #9858 .

### CODE REVIEW

File: typing/ctype.ml [hunk 1]
```diff
@@ -2514,7 +2516,10 @@ let complete_type_list ?(allow_absent=false) env nl1 lv2 mty2 nl2 tl2 =
         match Env.find_type_by_name lid env' with
         | (_, {type_arity = 0; type_kind = Type_abstract;
                type_private = Public; type_manifest = Some t2}) ->
-            (n, nondep_instance env' lv2 id2 t2) :: complete nl ntl2
+            begin match nondep_instance env' lv2 id2 t2 with
+            | exception Nondep_cannot_erase _ -> raise Exit
```
Comment (Octachron) [inline typing/ctype.ml]: Looking at the 4.07 code, a failure of `nondep_instance` was equivalent to `Exit` only in the case `allow_absent=false`. What is the reason for this change?

Comment (trefis) [reply typing/ctype.ml]: > What is the reason for this change?

My inattention.
Thanks for the catch.

**Review by Octachron — APPROVED** [review]
The code is correct: it restores the 4.07 error handling ... except for an accidental exponential blowup in the 4.07 version.

**Comment by Octachron** [comment]
For the record, the problematic behavior  in 4.07 could be triggered with
```ocaml

module type input = sig
  type a
  ...
  type z
end
module type output = sig
  type a
  ...
  type y
  type 'a z
end
let f (g: (module input
            with type a = int
             ...
             and type z = int)
      )
  =
  (g:> (module output with type a = int))
```

**Pull request merged**